### PR TITLE
[feature] Implement generic distance functions

### PR DESCRIPTION
**Pull request by lovasoa**
Closes https://github.com/dguo/strsim-rs/issues/29

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -79,7 +86,7 @@ pub fn jaro(a: &str, b: &str) -> f64 {
     let mut transpositions = 0.0;
     let mut b_match_index = 0;
 
-    for (i, a_char) in a.chars().enumerate() {
+    for (i, a_char) in a.into_iter().enumerate() {
```
Comment (dguo) [inline src/lib.rs]: Let's rename `a_char` and `b_char` to `a_elem` and `b_elem`.

File: src/lib.rs [hunk 2]
```diff
@@ -119,21 +126,41 @@ pub fn jaro(a: &str, b: &str) -> f64 {
...
+          Elem1: PartialEq<Elem2> {
+    let jaro_distance = jaro_generic(a, b);
 
     // Don't limit the length of the common prefix
-    let prefix_length = a.chars()
-                         .zip(b.chars())
-                         .take_while(|&(a_char, b_char)| a_char == b_char)
+    let prefix_length = a.into_iter()
+        .zip(b.into_iter())
+        .take_while(|&(ref a_char, ref b_char)| a_char == b_char)
```
Comment (dguo) [inline src/lib.rs]: Same as before: `char` to `elem`

File: src/lib.rs [hunk 3]
```diff
@@ -169,11 +209,11 @@ pub fn levenshtein(a: &str, b: &str) -> usize {
     let mut distance_a;
     let mut distance_b;
 
-    for (i, a_char) in a.chars().enumerate() {
+    for (i, a_char) in a.into_iter().enumerate() {
```
Comment (dguo) [inline src/lib.rs]: Same as above: `char` to `elem`

File: src/lib.rs [hunk 4]
```diff
@@ -291,7 +340,7 @@ pub fn damerau_levenshtein(a: &str, b: &str) -> usize {
         distances[1][j + 1] = j;
     }
 
-    let mut chars: HashMap<char, usize> = HashMap::new();
+    let mut chars: HashMap<Elem, usize> = HashMap::new();
```
Comment (dguo) [inline src/lib.rs]: `chars` to `elems`

File: src/lib.rs [hunk 5]
```diff
@@ -119,21 +126,41 @@ pub fn jaro(a: &str, b: &str) -> f64 {
     }
 }
 
-/// Like Jaro but gives a boost to strings that have a common prefix.
+struct StringWrapper<'a>(&'a str);
+
+impl<'a, 'b> IntoIterator for &'a StringWrapper<'b> {
```
Comment (dguo) [inline src/lib.rs]: Does this approach cause any sort of performance hit to using the string functions?

File: src/lib.rs [hunk 6]
```diff
@@ -262,15 +314,12 @@ pub fn osa_distance(a: &str, b: &str) -> usize {
...
 ///
-/// assert_eq!(2, damerau_levenshtein("ab", "bca"));
+/// assert_eq!(2, damerau_levenshtein_generic(&[1,2], &[2,3,1]));
 /// ```
-pub fn damerau_levenshtein(a: &str, b: &str) -> usize {
-    if a == b { return 0; }
-
-    let a_chars: Vec<char> = a.chars().collect();
-    let b_chars: Vec<char> = b.chars().collect();
+pub fn damerau_levenshtein_generic<Elem>(a_chars: &[Elem], b_chars: &[Elem]) -> usize
```
Comment (dguo) [inline src/lib.rs]: `chars` to `elems`

File: src/lib.rs (hunk 5 above)
Comment (lovasoa) [reply src/lib.rs]: No, this wrapper should disappear at compilation.

**Comment by lovasoa** [comment]
@dguo This is all fixed

**Comment by dguo** [comment]
I think you missed one part of my previous comment: 

> For consistency with the `normalized_*` functions, do you think it would make sense to call these `generic_*` rather than `*_generic`?

Do you think one way sounds more like idiomatic Rust than the other?

Also, for my own understanding, we don't have a good way to avoid having two versions of each function, right? I'm still a beginner at Rust, so I'm not sure what's possible. It seems like we can't just accept an iterator because for strings, we have to do `.chars()`. So we would either need a way to determine if the generic argument is a string, or we'd have to ask the user to call `.chars()`. Is that correct?

**Comment by lovasoa** [comment]
Unfortunately, strings [do not implement](https://stackoverflow.com/questions/49962611/why-does-str-not-implement-intoiterator/49963332) `IntoIterator`. So yes, we have to handle them separately. 
I'm sorry I missed the prefix thing, I'll change it.

**Comment by gchers** [comment]
Hi, thank you guys! Any updates on this?

**Comment by lovasoa** [comment]
I'm sorry I forgot about this. I just made the missing commit. @dguo, I think this is ready for merge.

**Comment by lovasoa** [comment]
Thank you @dguo !

**Comment by gchers** [comment]
thank you both!!

**Pull request merged**